src/hera/workflow.py

```diff
@@ -283,12 +278,14 @@ def __exit__(self, exc_type, exc_val, exc_tb) -> None:
 
     def add_task(self, t: Task) -> "Workflow":
         """Add a task to the workflow"""
-        add_task(self, t)
+        assert self.dag is not None, "A `DAG` must be defined when adding a task to a workflow"
```

Comment:
We may not need these now that we always set a value in `__init__`.

If this was to make `mypy` happy (ie: it thought `self.dag` was an `Optional[DAG]`), I think we can add a type hint to the original `self.dag = ` to be `self.dag: DAG = ...` (I'd expect mypy to infer that, but I think I've also seen issues with it's type narrowing when using ternaries).